src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1324,7 +1331,24 @@ public static bool HasIdentityConversion(TypeSymbol type1, TypeSymbol type2)
             Debug.Assert((object)type1 != null);
             Debug.Assert((object)type2 != null);
 
-            return type1.Equals(type2, TypeCompareKind.AllIgnoreOptions);
+            // PROTOTYPE(NullableReferenceTypes): Equals should return a value that indicates
+            // directly whether the types were equal or equal ignoring nullable modifiers.
+            if (!type1.Equals(type2, TypeCompareKind.AllIgnoreOptions))
+            {
+                return false;
+            }
+
+            if (!type1.Equals(type2, TypeCompareKind.AllIgnoreOptions | TypeCompareKind.CompareNullableModifiersForReferenceTypes | TypeCompareKind.UnknownNullableModifierMatchesAny))
+            {
+                if (useSiteDiagnostics == null)
+                {
+                    useSiteDiagnostics = new HashSet<DiagnosticInfo>();
+                }
+                // PROTOTYPE(NullableReferenceTypes): Need general WRN_NullabilityMismatch warning.
+                useSiteDiagnostics.Add(new CSDiagnosticInfo(ErrorCode.WRN_NullabilityMismatchInAssignment, type1, type2));
```

Comment:
>useSiteDiagnostics.Add(new CSDiagnosticInfo(ErrorCode.WRN_NullabilityMismatchInAssignment, type1, type2)); [](start = 16, length = 106)

Is it possible that we could end up with false positives? For example, we tried to check the identity conversion, but this wasn't used for final determination. We could even detect that the final conversion doesn't exist.